### PR TITLE
MVC Component Filter

### LINKED ISSUE #23: MVC component: Filter(Popup)
image

### PR DESCRIPTION
**Pull request by PavelKundenko**
### Added filter component 

image

image

image

### CODE REVIEW

**Review by Anastasiia0999 — APPROVED** [review]

File: app/components/filter/filter-controller.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { FilterView } from './filter-view.js';
```
Comment (so2niko) [inline app/components/filter/filter-controller.js:1]: Does we use spaces with {}?

If it's true, add this check to ES Lint, please. 
I've already seen code like this, but without spaces in Anastasia's PR.

File: app/components/filter/filter-model.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    const propertyValues = carsData.map((car) => car[key]);
+
+    return {
+      name: key,
+      values: [...new Set(propertyValues.flat(1))],
+    };
+  }).filter((category) => !this.ignoredCategories.includes(category.name));
+
+  extractFormData = (event) => [...event.target.querySelectorAll('input')].reduce((selectedProperties, formElement) => {
+    if (formElement.type !== 'submit') {
```
Comment (so2niko) [inline app/components/filter/filter-model.js]: Rewrite **3** if conditions. It is too much

File: app/components/filter/filter-view.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+export class FilterView {
+  rootNode = document.getElementById('root');
+
+  constructor(handleSubmit, formatName) {
+    this.handleSubmit = handleSubmit;
+    this.formatName = formatName;
+  }
+
+  render = (data) => {
+    this.rootNode.insertAdjacentHTML('afterbegin', `
```
Comment (so2niko) [inline app/components/filter/filter-view.js]: Cut this method to 2 methods

File: app/components/filter/filter-view.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        ${this.renderOptions(category)}
+      </fieldset>
+    </li>
+  `).join('');
+
+  renderOptions = (category) => {
+    const { name, values } = category;
+
+    if (typeof values[0] === 'number') {
+      return `
```
Comment (so2niko) [inline app/components/filter/filter-view.js]: Too much returns in method. Rewrite with 1 return

File: app/components/filter/filter-view.js [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+  removeActiveItem = () => {
+    const activeItem = this.modalContainer.querySelector('.filter__categories-item--active');
+    if (activeItem) {
+      activeItem.classList.remove('filter__categories-item--active');
+    }
+  };
+
+  showFilter = () => {
+    this.filterForm.parentNode.style.display = 'flex';
```
Comment (so2niko) [inline app/components/filter/filter-view.js]: Find it's parent in constructor or other method and link exact to it.

Do not use here parentNode

File: app/components/filter/filter-view.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    this.modalContainer.style.display = 'block';
+  };
+
+  hideFilter = () => {
+    this.removeActiveItem();
+    this.filterForm.parentNode.style.display = 'none';
```
Comment (so2niko) [inline app/components/filter/filter-view.js]: Second link to this element. Add property to view with link to this element

**Review by so2niko — CHANGES_REQUESTED** [review]

File: app/components/filter/filter-model.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  getCategoriesData = (carsData) => Object.keys(carsData[0]).map((key) => {
+    const propertyValues = carsData.map((car) => car[key]);
+
+    return {
+      name: key,
+      values: [...new Set(propertyValues.flat(1))],
+    };
+  }).filter((category) => !this.ignoredCategories.includes(category.name));
+
+  extractFormData = (event) => [...event.target.querySelectorAll('input')].reduce((selectedProperties, formElement) => {
```
Comment (so2niko) [inline app/components/filter/filter-model.js]: move this code to view

File: app/components/filter/filter-model.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+          ...selectedProperties[formElement.dataset.category] ?? [],
+          Number(formElement.value),
+        ],
+      };
+    } else {
+      return selectedProperties;
+    }
+  }, {});
+
+  filterData = (carsData, filterParams) => carsData
```
Comment (so2niko) [inline app/components/filter/filter-model.js:43]: too hard code.
6 if conditions and many `return false` construction.
Remaster this code

**Review by so2niko — CHANGES_REQUESTED** [review]

File: app/components/filter/filter-model.js (hunk 2 above)
Comment (so2niko) [reply app/components/filter/filter-model.js]: still there

Comment (so2niko) [reply app/components/filter/filter-model.js]: still there

File: app/components/filter/filter-view.js (hunk 3 above)
Comment (so2niko) [reply app/components/filter/filter-view.js]: 1 method - render
2 method - find elements and add event listeners to them

File: app/components/filter/filter-view.js (hunk 5 above)
Comment (so2niko) [reply app/components/filter/filter-view.js]: I don`t see changes

File: app/components/filter/filter-view.js (hunk 6 above)
Comment (so2niko) [reply app/components/filter/filter-view.js]: I don`t see changes

**Review by so2niko — CHANGES_REQUESTED** [review]

File: app/components/filter/filter-model.js (hunk 2 above)
Comment (PavelKundenko) [reply app/components/filter/filter-model.js]: It was fixed yesterday. Check latest commits, please.
image

File: app/components/filter/filter-view.js (hunk 5 above)
Comment (PavelKundenko) [reply app/components/filter/filter-view.js]: It was fixed yesterday, check latest commits, please

File: app/components/filter/filter-view.js (hunk 6 above)
Comment (PavelKundenko) [reply app/components/filter/filter-view.js]: Same as previous

File: app/components/filter/filter-model.js (hunk 8 above)
Comment (PavelKundenko) [reply app/components/filter/filter-model.js:43]: Check new version, please:
- flag was added to avoid early returns
- conditions structure became flat

**Review by so2niko — APPROVED** [review]

**Review by MozdolevskyiMaksym — APPROVED** [review]

**Pull request merged**